lucet-wasi-fuzz/src/main.rs

```diff
@@ -0,0 +1,267 @@
+    #[structopt(short = "n", long = "num-tests", default_value = "100")]
+    num_tests: usize,
+    #[structopt(long = "seed")]
+    seed: Option<Seed>,
+}
+
+fn main() {
+    lucet_runtime::lucet_internal_ensure_linked();
+
+    let config = Config::from_args();
+
+    if let Some(seed) = config.seed {
+        run_one_seed(seed);
+    } else {
+        run_many(config.num_tests);
+    }
+}
+
+fn run_one_seed(seed: Seed) {
+    match run_one(Some(seed)) {
+        Ok(TestResult::Passed) => println!("test passed"),
+        Ok(TestResult::Ignored) => println!("native build/execution failed"),
+        Ok(TestResult::Failed {
+            expected, actual, ..
+        }) => {
+            println!("test failed:\n");
+            println!("native: {}", String::from_utf8_lossy(&expected));
+            println!("lucet-wasi: {}", String::from_utf8_lossy(&actual));
+            std::process::exit(1);
+        }
+        Ok(TestResult::Errored { error }) | Err(error) => println!("test errored: {}", error),
+    }
+}
+
+fn run_many(num_tests: usize) {
+    let mut progress = progress::Bar::new();
+    progress.set_job_title(&format!("Running {} tests", num_tests));
+
+    let progress = Arc::new(Mutex::new(progress));
+    let num_finished = Arc::new(Mutex::new(0));
+
+    let res = (0..num_tests)
+        .into_par_iter()
+        .try_for_each(|_| match run_one(None) {
+            Ok(TestResult::Passed) | Ok(TestResult::Ignored) => {
+                let mut num_finished = num_finished.lock().unwrap();
+                *num_finished += 1;
+                let percentage = (*num_finished as f32 / num_tests as f32) * 100.0;
+                progress
+                    .lock()
+                    .unwrap()
+                    .reach_percent(percentage.floor() as i32);
+                Ok(())
+            }
+            Ok(fail) => Err(fail),
+            Err(error) => Err(TestResult::Errored { error }),
+        });
+
+    progress.lock().unwrap().jobs_done();
+
+    match res {
+        Err(TestResult::Failed {
+            seed,
+            expected,
+            actual,
+        }) => {
+            println!("test failed with seed {}\n", seed);
+            println!("native: {}", String::from_utf8_lossy(&expected));
+            println!("lucet-wasi: {}", String::from_utf8_lossy(&actual));
+            std::process::exit(1)
+        }
+        Err(TestResult::Errored { error }) => println!("test errored: {}", error),
+        Err(_) => unreachable!(),
+        Ok(()) => println!("all tests passed"),
+    }
+}
+
+fn gen_c<P: AsRef<Path>>(gen_c_path: P, seed: Seed) -> Result<(), Error> {
+    Command::new("csmith")
+        .arg("-s")
+        .arg(format!("{}", seed))
+        .arg("-o")
+        .arg(gen_c_path.as_ref())
+        .status()?;
+    Ok(())
+}
+
+fn run_native<P: AsRef<Path>>(tmpdir: &TempDir, gen_c_path: P) -> Result<Option<Vec<u8>>, Error> {
+    let gen_path = tmpdir.path().join("gen");
+
+    Command::new("cc")
```

Comment:
Wouldn't using the `cc` crate be better? It will improve portability, and the compiler and default options could be changed in a standard way.

Reply:
I didn't think `cc` was a general-purpose tool that could produce executables; I thought it was only for building libraries in `build.rs`?

Reply:
As long as `out_dir`, `target`, `opt_level` and `host` are defined, `cc` works anywhere.

Reply:
But this fuzzer is mainly for us, and it will only be used on one platform, so directly building the command line doesn't make a big difference anyway.

Reply:
Hmm, I tried this and still get a whole bunch of output like:

```
...
cargo:warning=                   ^
cargo:warning=/tmp/.tmp8uDaNl/gen.c:200:22: warning: 'g_2528' defined but not used [-Wunused-variable]
cargo:warning= static struct S1 ****g_2528 = (void*)0;
cargo:warning=                      ^
exit code: 0
AR_x86_64-unknown-linux = None
AR_x86_64_unknown_linux = None
HOST_AR = None
AR = None
running: "ar" "crs" "/tmp/.tmp8uDaNl/libgen.a" "/tmp/.tmp8uDaNl/gen.o"
exit code: 0
cargo:rustc-link-lib=static=gen
cargo:rustc-link-search=native=/tmp/.tmp8uDaNl
```

Reply:
`.flag_if_supported("-w")`

Reply:
The problem isn't the warnings, it's all the `cargo:*` lines for `build.rs` programs, which I discovered `build.cargo_metadata(false)` can disable. However it's still wanting to produce a library at the end, rather than an executable. I think I give up for now :slightly_smiling_face: